camel-core/src/main/java/org/apache/camel/processor/Throttler.java

```diff
@@ -235,20 +282,22 @@ protected void enqueuePermit(final ThrottlePermit permit, final Exchange exchang
     /**
      * Evaluates the maxRequestsPerPeriodExpression and adjusts the throttle rate up or down.
      */
-    protected void calculateAndSetMaxRequestsPerPeriod(final Exchange exchange) throws Exception {
+    protected void calculateAndSetMaxRequestsPerPeriod(DelayQueue<ThrottlePermit> delayQueue, final Exchange exchange, final Integer key) throws Exception {
         Integer newThrottle = maxRequestsPerPeriodExpression.evaluate(exchange, Integer.class);
 
         if (newThrottle != null && newThrottle < 0) {
             throw new IllegalStateException("The maximumRequestsPerPeriod must be a positive number, was: " + newThrottle);
         }
 
         synchronized (this) {
```

Comment:
As we separate throttleRate by the correlation id, we could use throttleRate as a lock instead of using this throttle object.

Reply:
CorrelationExpression may be even better?

Reply:
My understanding is that `correlationExpression` is always the same. Would the finer grained lock be the result of the evaluation then, so namely `key`?

Reply:
Yes i meant the evaluated correlationExpression. I agree